### PR TITLE
ARROW-6764: [C++] Create a readahead iterator

### PR DESCRIPTION
**Pull request by pitrou**
Convert the JSON reader to use it.  Performance is unchanged or a bit better.

### CODE REVIEW

**Comment by pitrou** [comment]
@bkietz

File: cpp/src/arrow/util/iterator.h [hunk 1]
```diff
@@ -312,4 +314,128 @@ Iterator<T> MakeFlattenIterator(Iterator<Iterator<T>> it) {
...
+
+    ARROW_RETURN_NOT_OK(queue_->Append(MakePromise()));
+
+    ARROW_RETURN_NOT_OK(it_promise->status_);
+    *out = std::move(it_promise->out_);
+    if (*out == IterationTraits<T>::End()) {
+      done_ = true;
+      // Shutdown the queue, but it may still have requested a couple more values
+      // in the meantime.
+      ARROW_RETURN_NOT_OK(queue_->Shutdown());
```
Comment (pitrou) [inline cpp/src/arrow/util/iterator.h]: Note to self: no need to actually shutdown the queue (`Shutdown(wait=true)` may take some time to return, and `Shutdown(wait=false)` is unsafe here).

File: cpp/src/arrow/util/iterator.cc [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+    return Status::OK();
+  }
+
+  Status Pump(std::function<std::unique_ptr<ReadaheadPromise>()> factory) {
+    std::unique_lock<std::mutex> lock(mutex_);
+    if (please_shutdown_) {
+      return Status::Invalid("Shutdown requested");
+    }
+    while (static_cast<int64_t>(done_.size() + todo_.size()) < max_readahead_) {
+      todo_.push_back(factory());
```
Comment (wesm) [inline cpp/src/arrow/util/iterator.cc]: It's possible that the initial pump could exhaust the iterator (for example, creating 5 tasks but only say 2 can be completed). I suppose it doesn't cause that much harm if the iterator's Next method can be called repeatedly after iteration has completed

File: cpp/src/arrow/util/iterator.cc [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+    DCHECK(!thread_.joinable());
+  }
+
+  Status Append(std::unique_ptr<ReadaheadPromise> promise) {
+    std::unique_lock<std::mutex> lock(mutex_);
+    if (please_shutdown_) {
+      return Status::Invalid("Shutdown requested");
+    }
+    todo_.push_back(std::move(promise));
+    if (static_cast<int64_t>(todo_.size()) == 1) {
```
Comment (wesm) [inline cpp/src/arrow/util/iterator.cc]: I assume if there were tasks in todo_ already then the worker will be woken up by design

**Review by wesm — COMMENTED** [review]
This is elegantly done and useful. Thanks for working on this. A small comment but otherwise LGTM

File: cpp/src/arrow/util/iterator.cc (hunk 2 above)
Comment (pitrou) [reply cpp/src/arrow/util/iterator.cc]: Yes, it shouldn't be a problem.

**Comment by pitrou** [comment]
@bkietz Would you like to take a look?

**Review by bkietz — APPROVED** [review]
This looks great

File: cpp/src/arrow/io/interfaces.cc [hunk 4]
```diff
@@ -39,6 +40,31 @@ FileInterface::~FileInterface() = default;
 
 Status FileInterface::Abort() { return Close(); }
 
+class InputStreamBlockIterator {
+ public:
+  InputStreamBlockIterator(std::shared_ptr<InputStream> stream, int64_t block_size)
+      : stream_(stream), block_size_(block_size) {}
+
+  Status Next(std::shared_ptr<Buffer>* out) {
+    if (done_) {
```
[the remark this answers is not in the document]
Comment (pitrou) [reply cpp/src/arrow/io/interfaces.cc]: The current code shows the intent better.

File: cpp/src/arrow/io/interfaces.cc [hunk 5]
```diff
@@ -51,6 +77,16 @@ Status InputStream::Peek(int64_t ARROW_ARG_UNUSED(nbytes),
 
 bool InputStream::supports_zero_copy() const { return false; }
 
+Status MakeInputStreamIterator(std::shared_ptr<InputStream> stream, int64_t block_size,
+                               Iterator<std::shared_ptr<Buffer>>* out) {
+  if (stream->closed()) {
+    return Status::Invalid("Cannot take iterator on closed stream");
+  }
+  DCHECK_GT(block_size, 0);
+  *out = Iterator<std::shared_ptr<Buffer>>(InputStreamBlockIterator(stream, block_size));
```
[the remark this answers is not in the document]
Comment (pitrou) [reply cpp/src/arrow/io/interfaces.cc]: My guess is that it would be less readable (the lambda would mutate outer variables).

**Comment by pitrou** [comment]
Rebased.

**Comment by pitrou** [comment]
AppVeyor: https://ci.appveyor.com/project/pitrou/arrow/builds/27968010

**Pull request closed without merging**